GraphQL over Internet blog post (#1182)
the-guild-org/website: website/pages/blog/graphql-over-internet.mdx

### [`graphql-http.com`](https://graphql-http.com)

Checking for compliance was never easier! [graphql-http.com](https://graphql-http.com) is a website
for swiftly auditing servers - even those that are running locally! It is always up-to-date with the
current spec.

Of course, security is very important - the website audits run exclusively in the browser, and there is
no tracking, data sharing, or external dependencies! Meaning you can
download the website on your computer and have a portable offline checker.

## [GraphQL over WebSocket](https://github.com/graphql/graphql-over-http/blob/main/rfcs/GraphQLOverWebSocket.md)

Before we talk about the current state of GraphQL WebSocket support, let's start with a brief
overview of the history. Originally there was this library
[`subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws) that was
built and maintained by [Apollo](https://www.apollographql.com/). It also had an
[accompanying protocol](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md).

> But as time passed, they've been able to spend less time on it, probably due to increase demand in
> other areas. The repo grew in issues, PRs were never reviewed - it eventually became unmaintained
> and therefore deprecated. But this library actually kick-started and pushed the evolution, it was a
> big inspiration and a great influence for the upcoming refresh!

Reviewer: ```suggestion
But as time passed, they haven't been able to spend as much time on it — probably due to increased demand in
other areas. The repo grew in issues, and PRs were never reviewed. Eventually, it became deprecated. However, this library actually kick-started and pushed the evolution. It was a
big inspiration and a great influence for an upcoming refresh!
```

client can send each other messages through the connection channel whenever they want. When used in
conjunction with GraphQL, it allows clients to subscribe and receive events from the server in
real-time. This makes it ideal for applications that require real-time updates and push
notifications.

The
[GraphQL over WebSocket protocol](https://github.com/graphql/graphql-over-http/blob/main/rfcs/GraphQLOverWebSocket.md)
leverages all pros of the WebSocket and gives shine to real-time GraphQL. Please note that it is
**not** backwards compatible with
[Apollo's protocol behind `subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md).

⚠️ Beware that the specification is currently an RFC and is open to changes!

### [Reference Implementation](https://github.com/enisdenjo/graphql-ws)

A specification goes best with a reference implementation. Say hello to
[`graphql-ws`](https://github.com/enisdenjo/graphql-ws). It is a zero-dependency server and client,
very simple and bare-bone. But don't be mistaken, it's very versatile and extendable!

> One key element people often struggle with WebSockets are re-connections and `graphql-ws` is all
> about stable connections. Silent retries with custom strategies are easy to use and extend.

Reviewer: (I tried to avoid double `with` — ie `struggle with with X`)
```suggestion
A common struggle when working with WebSockets is re-connections. `graphql-ws` is all
about stable connections. Silent retries with custom strategies are easy to use and extend.
```

Author: I guess it should be plural, as in:

> A common struggle when working with WebSockets _are_ re-connections.

or even better

> A common struggle when working with WebSockets is re-_connecting_.

What do you think?

Reviewer: Yeah, the second one sounds best! 

But I still think that it should be singular because `a struggle` is singular (and the right-hand side of the sentence doesn't matter since the `struggle` is the subject of this sentence)

Author: You're right, it should be "is". Sounded kinda' off... :D

I chose the second one anyways. Thanks!

| Latency       | Lower because of persisted TCP connection                                                                                                   | Can be higher sometimes because of multiplexing                                                                                                                                                             |
| Retries       | Needs to be developed in user-land                                                                                                          | Browsers have built-in retry mechanisms                                                                                                                                                                     |
| Support       | Browser native through [`WebSocket`](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket). No custom implementations for the browser | Browser native through [`EventSource`](https://developer.mozilla.org/en-US/docs/Web/API/EventSource). But, can be custom implemented on [`fetch()`](https://developer.mozilla.org/en-US/docs/Web/API/fetch) |

## Presentation

> Check out the video of the related presentation that happened at the

Reviewer: ```suggestion
Check out my presentation from
```

I think it's more clickable if it's obvious that the video is from the article's author

topics are discussed.

This is a joint effort, so everyone's opinion matters and everyone is welcome to contribute!

There is already a
[specification being standardized named GraphQL over HTTP](https://graphql.github.io/graphql-over-http/).
At the time of writing, it is in the proposal stage. It currently describes exclusively single
response operations, this is the main focus; so, no file uploads, no incremental deliveries, yet...

### [Reference Implementation](https://github.com/graphql/graphql-http)

To accompany the spec, the
[reference implementation `graphql-http`](https://github.com/graphql/graphql-http) is developed. It
is a zero-dependency server, client and audit suite.

The client is very simple to use and offers some great niceties - one of them being silent retries
of failed requests.

Beware that the library exclusively implements the
[current official GraphQL over HTTP spec](https://graphql.github.io/graphql-over-http/). It does so
in the clearest and most explicit manner so that others implementing their own server can quite
literally use [`graphql-http`](https://github.com/graphql/graphql-http) as a reference of the spec.

> The library is server agnostic and as such can run in any JavaScript environment. Additionally, it
> maintains the list of compliant servers and their reports - so please do PR your own implementation,
> you're more then welcome!

Reviewer: ```suggestion
The library is server agnostic and, as such, can run in any JavaScript environment. Additionally, it
maintains the list of compliant servers and their reports. You're more than welcome to send PRs with your own implementation!
```

### [`graphql-http.com`](https://graphql-http.com)

Checking for compliance was never easier! [graphql-http.com](https://graphql-http.com) is a website
for swiftly auditing servers - even those that are running locally! It is always up-to-date with the
current spec.

> Of course, security is very important - the website audits run exclusively in the browser, there is
> no tracking, no data sharing and there are no external dependencies either! Meaning, you can
> download the website on your computer and have a portable offline checker.

Reviewer: ```suggestion
Of course, security is very important - the website audits run exclusively in the browser, and there is
no tracking, data sharing, or external dependencies! Meaning you can
download the website on your computer and have a portable offline checker.
```
